python/tchannel/tornado/peer.py

```diff
@@ -551,22 +548,40 @@ def send(self, arg1, arg2, arg3,
         if request.is_streaming_request:
             request.ttl = 0
 
+        response = yield self.send_with_retry(
+            request, peer, attempt_times, retry_delay)
+
+        log.debug("Got response %s", response)
+        raise gen.Return(response)
+
+    @gen.coroutine
+    def _send(self, connection, req):
+        response_future = connection.send_request(req)
+        with timeout(response_future, req.ttl):
+            response = yield response_future
+
+        raise gen.Return(response)
+
+    @gen.coroutine
+    def send_with_retry(self, request, peer, attempt_times, retry_delay):
+        # black list to record all used peers, so they aren't chosen again.
+        blacklist = set()
+        response = None
+        connection = yield peer.connect()
         # mac number of times to retry 3.
         for num_of_attempt in range(attempt_times):
             try:
-                response = yield self._send(
-                    connection, request)
+                response = yield self._send(connection, request)
                 break
             except (ProtocolError, TimeoutError) as e:
+                # stop the outgoing request
                 request.set_exception(e)
-                if not request.should_retry_on_error(e):
+                if not request.should_retry_on_error(e) or (
+                        num_of_attempt == attempt_times - 1):
```

Comment:
Style nit: If the condition is becoming ungainly, move it into an aptly named variable.

    should_retry = (
        request.should_retry_on_error(e) and
        num_of_attempt <= ...
    )

    if not should_retry:
        ...